File: wwv_decoder.jl
using WAV
using DSP
using PyPlot

println("Libraries loaded...")

################################################################################

SAMPLE_RATE = None

################################################################################

function plot_filter(filter, hz=None)
    if hz == None
        freqs = 0:(SAMPLE_RATE/2)-1
    else
        freqs = hz
    end

    resp = 20*log10(abs(freqz(filter, freqs, SAMPLE_RATE)))
    plot(freqs, resp)
    readline()
end

################################################################################

function stream_wave_file(path, start=None, stop=None)
    global SAMPLE_RATE

    samples, SAMPLE_RATE = WAV.wavread(path)

    if start != None && stop == None
        samples = samples[int(start*SAMPLE_RATE)+1:end]
    elseif start != None && stop != None
        samples = samples[int(start*SAMPLE_RATE)+1:int(stop*SAMPLE_RATE)]
    end

    return samples
end

function stream_rectify(stream)
    Task(function ()
        for s in stream
            produce(abs(s))
        end
    end)
end

function stream_bandpass_filter_iir(stream, fLow, fHigh; realtime=false)
    bpf = convert(TFFilter, digitalfilter(Bandpass(fLow, fHigh; fs=SAMPLE_RATE), Butterworth(3)))
    #plot_filter(bpf, 0:1000)
    b, a = coefb(bpf), coefa(bpf)

    if !realtime
        return filt(b, a, collect(Float64, stream))
    else
        Task(function ()
            x = zeros(length(b))
            y = zeros(length(a)-1)

            for sample in stream
                # y[n] = 1/a_0 * ( (b_0*x[n] + b_1*x[n-1] + b_2*x[n-2] + ... b_N*x[n-N])
                #                   - (a_0*y[n-1] + a_1*y[n-2] + a_2*y[n-3] + ...) )
                x = cat(1, [sample], x[1:end-1])
                y_n = (dot(x, b) - dot(y, a[2:end])) / a[1]
                y = cat(1, [y_n], y[1:end-1])
                produce(y_n)
            end
        end)
    end
end

function stream_lowpass_filter_iir(stream, fC; realtime=false)
    bpf = convert(TFFilter, digitalfilter(Lowpass(fC; fs=SAMPLE_RATE), Butterworth(4)))
    #plot_filter(bpf, 0:1000)
    b, a = coefb(bpf), coefa(bpf)

    if !realtime
        return filt(b, a, collect(Float64, stream))
    else
        Task(function ()
            x = zeros(length(b))
            y = zeros(length(a)-1)

            for sample in stream
                # y[n] = 1/a_0 * ( (b_0*x[n] + b_1*x[n-1] + b_2*x[n-2] + ... b_N*x[n-N])
                #                   - (a_0*y[n-1] + a_1*y[n-2] + a_2*y[n-3] + ...) )
                x = cat(1, [sample], x[1:end-1])
                y_n = (dot(x, b) - dot(y, a[2:end])) / a[1]
                y = cat(1, [y_n], y[1:end-1])
                produce(y_n)
            end
        end)
    end
end

function stream_static_threshold(stream, threshold)
    Task(function ()
        for sample in stream
            if sample > threshold
                produce(1)
            else
                produce(0)
            end
        end
    end)
end

function stream_threshold(stream)
    state = []
    i = 0
    Task(function ()
        while length(state) < int(1.0*SAMPLE_RATE)
            state = cat(1, state, [float64(consume(stream))])
        end
        for sample in stream
            if i == 0 || i % 8000 == 0
                smin, smax = quantile(state, [0.15, 0.85])
                threshold = ((smax-smin)/2.0) + smin
            end

            if state[1] > 0.03 && state[1] > threshold
                produce(1)
            else
                produce(0)
            end

            state = cat(1, state[2:end], [sample])
            i += 1
        end
    end)
end

function stream_pulse_widths(stream)
    Task(function ()
        sample_number = 0
        state, width = (0, 0)

        for sample in stream
            sample_number += 1
            # 0 0
            if state == 0 && sample == 0
                state, width = (0, 0)
            # 0 1
            elseif state == 0 && sample == 1
                offset = sample_number
                state, width = (1, 1)
            # 1 1
            elseif state == 1 && sample == 1
                state, width = (1, width+1)
            # 1 0
            elseif state == 1 && sample == 0
                produce((offset/SAMPLE_RATE, width/SAMPLE_RATE))
                state, width = (0, 0)
            end
        end
    end)
end

function stream_filter_pulse_widths(stream)
    Task(function ()
        state = []

        for (offset, width) in stream
            if length(state) == 0
                if width > 125e-3
                    state = [(offset, width)]
                end
            else
                # Emit the collected pulse if this one is 900ms past the beginning
                # of the collection
                if abs(offset - state[1][1]) > 900e-3
                    produce((state[1][1], sum(map((t) -> t[2], state))))
                    state = [(offset, width)]
                # Add this pulse to our collection if it's less than 800ms from the
                # end of the last pulse in our collection
                elseif abs(offset - state[end][1]) < 800e-3
                    state = cat(1, state, [(offset, width)])
                end
            end
        end

        # Flush the last pulse collection when input stream is terminated
        if length(state) > 0
            produce((state[1][1], sum(map((t) -> t[2], state))))
        end
    end)
end

function stream_pulse_widths_to_symbols(stream)
    Task(function ()
        # approximately equal means actual vs. +/-25% of expected
        approx_equal = (actual, expected) -> abs(actual - expected) < 0.25*expected

        for (offset, width) in stream
            # 800ms for a marker
            if approx_equal(width, 800e-3)
                produce((offset, "M"))
            # 500ms for a 1 bit
            elseif approx_equal(width, 500e-3)
                produce((offset, 1))
            # 200ms for a 0 bit
            elseif approx_equal(width, 200e-3)
                produce((offset, 0))
            # Invalid bit
            else
                produce((offset, "I"))
            end
        end
    end)
end

function stream_symbols_to_frame(stream)
    compare = (expected, actual) -> (expected == "B" && (actual == 0 || actual == 1)) || actual == expected
    template = [     0 , "B", "B", "B", "B", "B", "B",  0 , "M",
                    "B", "B", "B", "B",  0 , "B", "B", "B",  0 , "M",
                    "B", "B", "B", "B",  0 , "B", "B",  0 ,  0 , "M",
                    "B", "B", "B", "B",  0 , "B", "B", "B", "B", "M",
                    "B", "B",  0 ,  0 ,  0 ,  0 ,  0 , "B",  0 , "M", # FIXME
                    "B", "B", "B", "B", "B", "B", "B", "B", "B", "M"    ]
    state = repmat([None], 59)

    Task(function ()
        for (offset, symbol) in stream
            println((offset, symbol))
            state = cat(1, state[2:end], [(offset, symbol)])

            if !(None in state)
                # Check that the symbols are each 1.0s second +/-100ms apart
                offsets = collect(Float64, map((t) -> t[1], state))
                if !all(abs(diff(offsets) - 1.0) .< 100e-3)
                    continue
                end

                # Check for no invalid symbols
                symbols = map((t) -> t[2], state)
                if "I" in symbols
                    continue
                end

                # Check that the symbols match the template
                matches = map((i) -> compare(template[i], state[i][2]), 1:length(template))
                if !all(matches)
                    continue
                end

                # Emit the symbols of the valid frame
                println("Valid frame found!")
                produce(symbols)

                # Reset the state
                state = repmat([None], 59)
            end
        end
    end)
end

type WWVRecord
    DST1::Bool
    LSW::Bool
    Year::Int
    Minutes::Int
    Hours::Int
    DayOfYear::Int
    DUT1::Bool
    DST2::Bool
    UT1Corr::Float64
end

function stream_frame_to_wwv_record(stream)

    Task(function ()
        for frame in stream
            dst1 = bool(frame[2])
            lsw = bool(frame[3])
            year = 1*frame[4] + 2*frame[5] + 4*frame[6] + 8*frame[7] + 10*frame[51] + 20*frame[52] + 40*frame[53] + 80*frame[54]
            minutes = 1*frame[10] + 2*frame[11] + 4*frame[12] + 8*frame[13] + 10*frame[15] + 20*frame[16] + 40*frame[17]
            hours = 1*frame[20] + 2*frame[21] + 4*frame[22] + 8*frame[23] + 10*frame[25] + 20*frame[26]
            day_of_year = 1*frame[30] + 2*frame[31] + 4*frame[32] + 8*frame[33] + 10*frame[35] + 20*frame[36] + 40*frame[37] + 80*frame[38] + 100*frame[40] + 200*frame[41]
            dut1 = bool(frame[50])
            dst2 = bool(frame[55])
            ut1_corr = 0.1*frame[56] + 0.2*frame[57] + 0.3*frame[58]

            produce(WWVRecord(dst1, lsw, year, minutes, hours, day_of_year, dut1, dst2, ut1_corr))
        end
    end)
end

function stream_print_wwv_record(stream)
    for record in stream
        println(record)
    end
end

function stream_plot(stream, n=None)
    if n == None
        data = collect(Float64, stream)
    else
        # FIXME
    end

    plot(0:length(data)-1, data)
    readline()
end

################################################################################

if length(ARGS) == 0
    @printf "Usage: <WWV recording wave file> [start] [stop]\n"
    exit(1)
end

println("Loading WAV file...")
if length(ARGS) == 1
    s0 = stream_wave_file(ARGS[1])
elseif length(ARGS) == 2
    s0 = stream_wave_file(ARGS[1], float(ARGS[2]))
elseif length(ARGS) >= 3
    s0 = stream_wave_file(ARGS[1], float(ARGS[2]), float(ARGS[3]))
end

println("Starting processing...")
s1 = stream_bandpass_filter_iir(s0, 95.0, 105.0; realtime=true)
s2 = stream_rectify(s1)
s3 = stream_lowpass_filter_iir(s2, 5.0; realtime=true)
s4 = stream_static_threshold(s3, 0.03)
s5 = stream_pulse_widths(s4)
s6 = stream_filter_pulse_widths(s5)
s7 = stream_pulse_widths_to_symbols(s6)
s8 = stream_symbols_to_frame(s7)
s9 = stream_frame_to_wwv_record(s8)
stream_print_wwv_record(s9)

File: wwv_decoder.py
import sys
import time
import collections
import numpy
import numpy.fft
import scipy.io.wavfile
import scipy.signal
import matplotlib.pyplot as plt

################################################################################

def plot_dft_samples(samples, sample_rate, title=""):
    samples_dft_mag = 20*numpy.log10(numpy.abs(numpy.fft.fft(samples)))
    samples_freqs = numpy.fft.fftfreq(len(samples_dft_mag), d=1.0/sample_rate)

    plt.plot(samples_freqs, samples_dft_mag)
    plt.xlim([-9000, 9000])
    plt.ylabel('Amplitude (Log)')
    plt.xlabel('Frequency (Hz)')
    plt.title(title)
    plt.show()

def plot_filter(b, a, sample_rate, freqs=None, title=""):
    if freqs is None:
        w, h = scipy.signal.freqz(b, a)
    else:
        w, h = scipy.signal.freqz(b, a, (2*numpy.pi*numpy.array(freqs))/sample_rate)

    plt.plot((sample_rate * w)/(2*numpy.pi), 20*numpy.log10(abs(h)))
    plt.ylabel('Amplitude (dB)')
    plt.xlabel('Frequency (Hz)')
    plt.title(title)
    plt.show()

def timed(message):
    def wrap(f):
        def wrapped_f(*args, **kwargs):
            sys.stdout.write(message + "\n")
            sys.stdout.flush()
            tic = time.time()
            rets = f(*args, **kwargs)
            toc = time.time()
            sys.stdout.write(" "*50 + "%.3f sec\n" % (toc-tic))
            return rets
        return wrapped_f
    return wrap

################################################################################

SAMPLE_RATE = None
THRESHOLD = None

################################################################################

@timed("Reading wave file...")
def block_wave_file(path, start=None, stop=None):
    global SAMPLE_RATE

    (SAMPLE_RATE, samples) = scipy.io.wavfile.read(path, mmap=True)

    if start is not None and stop is None:
        samples = samples[start*SAMPLE_RATE:]
    elif start is not None and stop is not None:
        samples = samples[start*SAMPLE_RATE:stop*SAMPLE_RATE]

    return samples

@timed("Bandpass filtering...")
def block_bandpass_filter_iir(samples, fLow, fHigh):
    b,a = scipy.signal.butter(3, [(2*fLow)/(SAMPLE_RATE), (2*fHigh)/(SAMPLE_RATE)], btype='bandpass')
    #plot_filter(b, a, SAMPLE_RATE, range(1000))
    return scipy.signal.lfilter(b, a, samples)

@timed("Rectifying...")
def block_rectify(samples):
    return numpy.abs(samples)

@timed("Low pass filtering...")
def block_lowpass_filter_iir(samples, fC):
    b, a = scipy.signal.butter(4, (2*fC)/SAMPLE_RATE)
    #plot_filter(b, a, SAMPLE_RATE, range(1000))
    return scipy.signal.lfilter(b, a, samples)

@timed("Finding threshold...")
def block_find_threshold(samples):
    global THRESHOLD

    counts, bins = numpy.histogram(samples, bins=100)
    smode = bins[numpy.argmax(counts)]

    smin = numpy.min(samples)
    smax = numpy.max(samples)
    smean = numpy.mean(samples)
    stdev = numpy.std(samples)

    THRESHOLD = smode + stdev

    print "    min %.2f  max %.2f  mean %.2f  stdev %.2f" % (smin, smax, smean, stdev)
    print "    approx. mode %.2f  threshold %.2f" % (smode, THRESHOLD)

    #plt.hist(samples, bins=100)
    #plt.axvline(THRESHOLD, color='r', linestyle='dashed')
    #plt.show()

    return samples

@timed("Thresholding...")
def block_threshold(samples):
    samples = numpy.copy(samples)

    idx_above = samples > THRESHOLD
    idx_below = samples < THRESHOLD
    samples[idx_above] = 1
    samples[idx_below] = 0

    return samples

@timed("Converting samples to pulse widths...")
def block_pulse_widths(samples):
    widths = []

    markers = numpy.diff(samples)
    starts, = numpy.where(markers > 0)
    stops, = numpy.where(markers < 0)
    widths = (stops - starts)/float(SAMPLE_RATE)
    offsets = (starts + 1)/float(SAMPLE_RATE)

    pulses = zip(offsets, widths)

    return pulses

@timed("Filtering pulse widths...")
def block_filter_pulse_widths(samples):
    state = []

    for (offset, width) in samples:
        if len(state) == 0:
            if width > 125e-3:
                state = [(offset, width)]
        else:
            # Emit the collected pulse if this one is 900ms past the beginning
            # of the collection
            if abs(offset - state[0][0]) > 900e-3:
                yield (state[0][0], sum([w for (_,w) in state]))
                state = [(offset, width)]
            # Add this pulse to our collection if it's less than 800ms from the
            # end of the last pulse in our collection
            elif abs(offset - state[-1][0]) < 800e-3:
                state.append((offset, width))

    # Flush the last pulse collection when input samples is terminated
    if len(state) > 0:
        yield (state[0][0], sum([w for (_,w) in state]))

@timed("Converting pulse widths to symbols...")
def block_pulse_widths_to_symbols(samples):
    # approximately equal means actual is within +/- 25% of expected
    approx_equal = lambda actual, expected: abs(actual - expected) < 0.25*expected

    for (offset, width) in samples:
        # 800ms for a marker
        if approx_equal(width, 800e-3):
            yield (offset, "M")
        # 500ms for a 1 bit
        elif approx_equal(width, 500e-3):
            yield (offset, 1)
        # 200ms for a 0 bit
        elif approx_equal(width, 200e-3):
            yield (offset, 0)
        # Invalid bit
        else:
            yield (offset, "I")

@timed("Converting symbols to frame...")
def block_symbols_to_frame(samples):
    template = [     0 , 'B', 'B', 'B', 'B', 'B', 'B',  0 , 'M',
                    'B', 'B', 'B', 'B',  0 , 'B', 'B', 'B',  0 , 'M',
                    'B', 'B', 'B', 'B',  0 , 'B', 'B',  0 ,  0 , 'M',
                    'B', 'B', 'B', 'B',  0 , 'B', 'B', 'B', 'B', 'M',
                    'B', 'B',  0 ,  0 ,  0 ,  0 ,  0 , 'B',  0 , 'M', # FIXME
                    'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M'    ]
    state = [None]*59

    for (offset, symbol) in samples:
        print (offset, symbol)
        state = state[1:] + [(offset, symbol)]
        if None not in state:
            # Check that the symbols are each 1.0s second +/-100ms apart
            offsets = [offset for (offset, _) in state]
            if not numpy.all(numpy.abs(numpy.diff(offsets) - 1.0) < 100e-3):
                continue

            # Check for no invalid symbols
            symbols = [symbol for (_, symbol) in state]
            if "I" in symbols:
                continue

            # Check that the symbols match the template
            compare = lambda expected, actual: (expected == 'B' and (actual == 0 or actual == 1)) or actual == expected
            matches = [ compare(template[i], state[i][1]) for i in range(len(template)) ]
            if not numpy.all(matches):
                continue

            # Emit the symbols of the valid frame
            print "Valid frame found!"
            yield symbols

            # Reset the state
            state = [None]*59

def block_frame_to_wwv_record(samples):
    WWVRecord = collections.namedtuple('WWVRecord', ['DST1', 'LSW', 'Year', 'Minutes', 'Hours', 'Day_of_year', 'DUT1', 'DST2', 'UT1_Corr'])

    for frame in samples:
        dst1 = bool(frame[1])
        lsw = bool(frame[2])
        year = 1*frame[3] + 2*frame[4] + 4*frame[5] + 8*frame[6] + 10*frame[50] + 20*frame[51] + 40*frame[52] + 80*frame[53]
        minutes = 1*frame[9] + 2*frame[10] + 4*frame[11] + 8*frame[12] + 10*frame[14] + 20*frame[15] + 40*frame[16]
        hours = 1*frame[19] + 2*frame[20] + 4*frame[21] + 8*frame[22] + 10*frame[24] + 20*frame[25]
        day_of_year = 1*frame[29] + 2*frame[30] + 4*frame[31] + 8*frame[32] + 10*frame[34] + 20*frame[35] + 40*frame[36] + 80*frame[37] + 100*frame[39] + 200*frame[40]
        dut1 = bool(frame[49])
        dst2 = bool(frame[54])
        ut1_corr = 0.1*frame[55] + 0.2*frame[56] + 0.3*frame[57]

        yield WWVRecord(dst1, lsw, year, minutes, hours, day_of_year, dut1, dst2, ut1_corr)

def block_print_wwv_record(samples):
    for record in samples:
        print record

def block_plot(samples, n=None, title=""):
    plt.plot(samples[0:n])
    plt.ylabel('Value')
    plt.xlabel('Time (sample number)')
    plt.title(title)
    plt.show()

################################################################################

if len(sys.argv) < 2:
    print "Usage: %s <recorded WWV wave file> [start] [stop]" % sys.argv[0]
    sys.exit(1)
elif len(sys.argv) == 2:
    samples = block_wave_file(sys.argv[1])
elif len(sys.argv) == 3:
    samples = block_wave_file(sys.argv[1], int(sys.argv[2]))
elif len(sys.argv) >= 4:
    samples = block_wave_file(sys.argv[1], int(sys.argv[2]), int(sys.argv[3]))

samples = block_bandpass_filter_iir(samples, 95.0, 105.0)
samples = block_rectify(samples)
samples = block_lowpass_filter_iir(samples, 5.0)
samples = block_find_threshold(samples)
samples = block_threshold(samples)
samples = block_pulse_widths(samples)
samples = block_filter_pulse_widths(samples)
samples = block_pulse_widths_to_symbols(samples)
samples  = block_symbols_to_frame(samples)
samples = block_frame_to_wwv_record(samples)
block_print_wwv_record(samples)

